### ISSUE TITLE
reCaptcha Audio Chalenge - Flagged as bot

### ISSUE THREAD

**Issue by rodrigovitachi** [cefsharp/CefSharp#1999]
I keep a school project for blind people, which allows them to use our intranet and interact with internal and external platforms.

The problem is that recently, the CefSharp started to be flagged as bot by the Google reCaptcha, when using the audio chalenge.

It doesn't play the audio chalenge, then it's necessary to click the download button in order to download the mp3 chalenge file. And that is when CefSharp is flagged as bot.

If i try it with my regular chrome browser, it works, and most important, if i try it with the core CefClient application, it also works, the mp3 file comes with the captcha chalenge.

Current build ( using  Visual Studio NuGet )
CEF 3.2883.1552.g88ff29a
Chromium 55.0.2883.87

Using Winforms version, x64 Windows 10 ( developer machine ) but also tested with many other machines using Windows 8, Windows Server, etc. Same results with all machines.

In order to reproduce the problem, it's only necessary to:
- Add a basic CefSharp instance
- Add the download handler code
- Load any url that contains the reCaptcha chalenge ( i.e: https://www.google.com/recaptcha/api2/demo )
- Click the captcha, open the audio chalenge, and then download the mp3 chalenge file.

DownloadHandler code ( got it from the forum ):
```c#
public class DownloadHandler : IDownloadHandler
    {
        ChromiumWebBrowser browser;
        public event EventHandler<DownloadItem> OnBeforeDownloadFired;
        public event EventHandler<DownloadItem> OnDownloadUpdatedFired;


        public void OnBeforeDownload(IBrowser browser, DownloadItem downloadItem, IBeforeDownloadCallback callback)
        {
            var handler = OnBeforeDownloadFired;
            if (handler != null)
            {
                handler(this, downloadItem);
            }

            if (!callback.IsDisposed)
            {
                using (callback)
                {
                    callback.Continue("C:\\Downloads\\teste.mp3", showDialog: false);
                }
            }
        }

        public void OnDownloadUpdated(IBrowser browser, DownloadItem downloadItem, IDownloadItemCallback callback)
        {
            var handler = OnDownloadUpdatedFired;
            if (handler != null)
            {
                handler(this, downloadItem);
            }

            if (downloadItem.IsComplete)
            {
                Console.WriteLine(downloadItem.Url);
            }
        }
    }
```

After doing it and after the download is completed, you should play the mp3 and be able to hear the chalenge, but what happens instead is that what you will hear is Google telling you that you are performing automatic tasks and because of that, the captcha is blocked.

Just to be clear again, this problem doesn't happen with the minimal cefclient software, it only happens when trying to use CefSharp as web client.

**Comment by amaitland** [comment]
> It doesn't play the audio chalenge, then it's necessary to click the download button in order to download the mp3 chalenge file. And that is when CefSharp is flagged as bot.

For clarification `MP3` is a proprietary codec, so `CefSharp` cannot play it natively due to licensing. See #1479 for background. 

> Just to be clear again, this problem doesn't happen with the minimal cefclient software, it only happens when trying to use CefSharp as web client.

I tested with `cef_binary_3.2987.1597.gffc5773_windows32_client` a couple of times, it's hit and miss whether it works. So it's not `CefSharp` specific. There are likely other factors that come into play, like `IP Address`, number of requests in a given time period.

If I think of anything that's helpful I'll let you know.

**Comment by rodrigovitachi** [comment]
> There are likely other factors that come into play, like IP Address, number of requests in a given time period.

Yes, i agree. Just to clarify, like i said, i tested with other browsers, such as my regular chrome, and also cefclient, and both works, then it's not related to my machine, ip, or number of requests, since they are the same machine and ip.

**Comment by amaitland** [comment]
> Yes, i agree. Just to clarify, like i said, i tested with other browsers, such as my regular chrome, and also cefclient, and both works

I've just tested in `FireFox` using https://www.google.com/recaptcha/api2/demo and I'm getting the same error audio file playing. Now I'm getting the same thing in `Chrome`, so I'm guessing too many requests from my `IP` has triggered a blacklist.

**Comment by amaitland** [comment]
For reference there is a report on [ceforum](http://magpcss.org/ceforum/viewtopic.php?f=6&t=14907&p=33863) suggesting recpatcha is many times harder when using `CEF`. 

I do think it's worth someone starting a conversation on http://magpcss.org/ceforum/viewforum.php?f=6
Possibly someone else has some ideas or has investigated further.

**Comment by merceyz** [comment]
I had a similar problem with the normal captchas (not tested on audio) and found that using the `simpleSubProcess` worked much better.

Long story short, commenting out this line helped a lot.
https://github.com/cefsharp/CefSharp/blob/master/CefSharp.BrowserSubprocess.Core/CefAppUnmanagedWrapper.cpp#L488

**Comment by joe-williams-cccu** [comment]
Some of my users have reported this issue, and I also have experienced this with `reCaptures` when using https://www.google.com/recaptcha/api2/demo. The `reCaptcha` was unsolvable for me and audio capture didn't work, either (although reading here audio capture may use proprietary codecs). However, when I tested on a different website that also implements Google's `reCaptcha`, I was **successful** in solving. 

`CefSharp 55.0.0` `x86` `WinForms` `Windows 10 64-bit` (also occurred on a `Windows 7 64-bit` machine)

**Comment by rodrigovitachi** [comment]
> However, when I tested on a different website that also implements Google's reCaptcha (image-based, again)

Yes, when using the image chalenge, its usually a lot harder than the normal, but it is possible to solve most of the times. The real problem is in the audio captcha, which is what i need, considering the visual condition of my users.

**Comment by rodrigovitachi** [comment]
@merceyz Thank you for your reference. Unfortunately it didn't help with reCaptcha audio chalenge, but i appreciate your support.

**Comment by merceyz** [comment]
Closing due to lack of feedback

**Comment by dquirino** [comment]
Hi,
When I use simple sample, google recaptha works.
When I use tabBrowser, google recaptcha do not works.
I had tried something and see that, if I change SubProcess to app name, google recaptcha works normal but I can not do this because EvaluateScriptAsync use SubProcess and it freeze if not use SubProcess.

I created this issue but the reader read and does not understood the question.

cefsharp/CefSharp#2473

Someone to help me?

**Comment by amaitland** [comment]
> Long story short, commenting out this line helped a lot.
> https://github.com/cefsharp/CefSharp/blob/master/CefSharp.BrowserSubprocess.Core/CefAppUnmanagedWrapper.cpp#L488

Assuming the reference was to `CefRegisterExtension` which is used for `Promise Creation`. https://github.com/cefsharp/CefSharp/blob/cefsharp/57/CefSharp.BrowserSubprocess.Core/CefAppUnmanagedWrapper.cpp#L488

I've removed this in https://github.com/cefsharp/CefSharp/commit/7450d0d51b1d8bfbe7f3aedc7ce8fe3c96e83c03 which is an attempt to resolve #2432 (which actually seems to be resolved by upgrading to a newer `CEF` version. ).

The `Audio Recaptch` requires `MP3` to play the audio which is now supported. So a very quick test and I was able to resolve an `Audio Recaptcha` using https://github.com/cefsharp/CefSharp/commit/7450d0d51b1d8bfbe7f3aedc7ce8fe3c96e83c03

This is a very quick test and I make no guarantees that it works all the time or will work in the future. It's just a happy coincidence that it appears to be working now. At this point the changes will land in version `69` (No time frame for this release)
